### ISSUE TITLE
hypothetical importance scores

### ISSUE THREAD

**Issue by dmaloneynygc** [kundajelab/tfmodisco#5]
Can you very briefly explain the hypothetical importance scores that accompanies importance scores in the TF MoDISco TAL GATA example? Are these the sequences with randomized importance scores?

**Comment by dmaloneynygc** [comment]
Thanks for such a timely response. Im using deeplift for importance scores. 

What is the dim of input_references_list[0]? 
i"m guessing it should be input_data_list [0].shape == input_references_list[0].shape? even in the instance we used more than one reference when getting importance scores?

**Comment by dmaloneynygc** [comment]
Yep. Thats what I was referring to -- get_shuffle_seq_ref_function() is what I used to get importance scores. 

This helps. Thanks

**Comment by sholderbach** [comment]
Hi @AvantiShri,
I am currently working with a model for which working with DeepLift would require some additional implementation work. So far I am using integrated gradients to create my attribution/importance values. What would be your intuition for a sound hypothetical importance and importance in this case? Following your suggestions for grad*input I would assume as hypothetical importance the averaged gradients along the path (zero centered)

**Comment by AvantiShri** [comment]
Hi @sholderbach,

Yes, the average gradient is a fine choice. Some additional nuance could be applied to take the value of the reference into account, but I expect simply using the average gradient will work.

Incidentally, consider looking at the implementation in the [DeepExplain](https://github.com/marcoancona/DeepExplain) repo if you want a faster alternative to integrated gradients (the authors of that paper found that DeepLIFT with the rescale rule (which is what they implemented) had very high correlation with integrated gradients, making it a good faster approximation).

If you are using a reference-based scoring method (like deeplift or integrated gradients) with sequence data, I do recommend taking special care with the choice of reference. Our best “recipe” for a reference for sequence input is to take the average over multiple dinucleotide-shuffled versions of the original sequence (this is illustrated in the deeplift repo’s example genomics notebook). If you are using a single reference per sequence, I suggest checking the output of the model on that reference; know that the contribution scores you get from integrated gradients will explain the difference in output from that reference output, so make sure the reference output is something appropriate for the question you are asking.

Incidentally, what is the additional implementation that DeepLIFT would need? Just asking because it’s helpful for me to get a sense of which features are in demand from the community.

**Comment by sholderbach** [comment]
Thank you very much for the quick and thorough response!
What kind of detrimental effects do you observe with a bad/naive reference sequence, when it comes to motif discovery? 
Regarding DeepLIFT, I had reimplemented a reverse complement convolution op myself and was unsure how to interface that to DeepLIFT. As I am working on some dynamic filtering components I have a non sequential topology/ have implemented some stuff with basic tensorflow ops. I was unsure to which extent this is compatible with using DeepLIFT.

**Comment by AvantiShri** [comment]
If you've implemented things with basic tensorflow ops, you're probably best off using integrated grads unless the compute time becomes a hassle. My understanding is that the DeepExplain repo implemented the DeepLIFT rescale rule by overriding backprop operators in the tensorflow ops, so that might be a good next step if you need a faster alternative to integrated gradients.

The extent to which a naive reference hurts the scores varies from model to model. Sometimes it behaves fine, and sometimes it gives very odd results. To appreciate why, note that when you have a single flat reference across the entire sequence (e.g. a reference of ACGT frequencies), every position in the convolutional layer is receiving the same input. Sometimes, a convolutional neuron may fire in response to this letter frequencies input, meaning that the neuron will fire with the same activation across the whole sequence - something that would never occur on real sequences. As a result of having a convolutional neuron firing at the same activation level across the entire sequence, the higher layers may respond in unexpected ways, occasionally resulting in reference activations that are inappropriate for the question you are trying to ask. Of course, sometimes you get lucky and the importance scores are well-behaved despite using a naive reference.

The best thing to do to assess the quality of a reference is (1) look at the prediction the model gives on the reference, as discussed above, and make sure it's consistent with what you expect, and (2) visualize the importance scores on a few sequences where you have an idea of which patterns you expect to see. If the reference is tanking, it will be apparent from viewing the scores.

**Comment by sholderbach** [comment]
I checked the source of DeepExplain, if necessary I will give it a shot on that. 
As I am currently working with methylation values I definetely have to take the plausibility of the reference into account. What I noticed so far is that sometimes small spikes of ~2 nucleotides appear in the importance with greater magnitude compared to the expected motif like patterns.

**Comment by AvantiShri** [comment]
@dmaloneynygc I put together a notebook demonstrating how to get the hypothetical importance scores using DeepLIFT: https://github.com/kundajelab/tfmodisco/blob/e1dcf39fe968e087315f63ec002d842abe3072c3/examples/simulated_tf_binding/Generate%20Importance%20Scores.ipynb

The corresponding TF-MoDISco run is here: https://github.com/kundajelab/tfmodisco/blob/e1dcf39fe968e087315f63ec002d842abe3072c3/examples/simulated_tf_binding/TF%20MoDISco%20TAL%20GATA.ipynb

FYI, I have also updated the seqlet thresholds to use a null based on a Laplace distribution (the previous version is still available under Branch-->Tags-->v0.2-alpha).
